### DISCUSSION TITLE
Can shapes be shown on paper in joint js without drag and drop? Can we click on shapes and show them on paper?

### DISCUSSION THREAD

**Discussion by nikita-fxbytes** [clientIO/joint#2509] [Q&A]
### Introduction

https://www.jointjs.com/demos/bpmn
In this, instead of dragging and dropping the shape, you can show it on the paper by just clicking on the shape.

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS+

**my code is**

```jsx
'use client'
import React, { useEffect, useRef, useState } from 'react';
import { dia, ui, shapes, elementTools,linkTools, layout } from '@clientio/rappid';
import _ from 'lodash';
import { toolList } from './toolbar';
import { examples } from './example';
import { inputss } from './inspector';
const Bpmne = () => {
    const toolbarContainer: React.RefObject<HTMLDivElement> = useRef(null);
    const stencilContainer: React.RefObject<HTMLDivElement> = useRef(null);
    const paperContainer: React.RefObject<HTMLDivElement> = useRef(null);


    const [droppedElements, setDroppedElements] = useState<any>([]); // State to hold dropped elements


    const getEl = (e:any) => {
      var inputs:any = inputss;
      var cell = e.model;
      var type = cell.get('type');
      ui.Inspector.create('#inspector-container', {
        cell: cell,
        inputs: inputs[type],
        groups: {
          general: { label: type, index: 1 },
          appearance: { index: 2 },
          defaults: { index: 3 }
        }
      });
    }
  useEffect(() => {
    if (
      paperContainer.current &&
      toolbarContainer.current &&
      stencilContainer.current
    ) {
    var example = examples;
    var inputs:any = inputss;
    var toolbarConfig = toolList;
    var bpmn2:any = shapes.bpmn2;
    var BPMNTypes = {
      Pool: 'bpmn2.HeaderedPool',
      Group: 'bpmn2.Group',
      Activity: 'bpmn2.Activity',
      Event: 'bpmn2.Event',
      Gateway: 'bpmn2.Gateway',
      DataObject: 'bpmn2.DataObject',
      DataStore:'bpmn2.DataStore',
      DataAssociation: 'bpmn2.DataAssociation',
      Flow: 'bpmn2.Flow',
      Conversation: 'bpmn2.Conversation',
      ConversationLink: 'bpmn2.ConversationLink',
      Annotation: 'bpmn2.Annotation',
      AnnotationLink: 'bpmn2.AnnotationLink',
      // Legacy BPMN Shape
      Choreography: 'bpmn.Choreography'
    };
    var graph = new dia.Graph({ type: 'bpmn' }, { cellNamespace: shapes });
    var commandManager = new dia.CommandManager({ graph: graph });
    var keyboard = new ui.Keyboard();
  
  
    var paper = new dia.Paper({
      width: 1500,
      height: 500,
      model: graph,
      gridSize: 5,
      async: true,
      sorting: dia.Paper.sorting.APPROX,
      interactive: { linkMove: false },
      snapLabels: true,
      cellViewNamespace: shapes,
      clickThreshold: 10,
      // Connections
      defaultLink: function () {
        return new bpmn2.Flow({
          attrs: {
              line: {
                  flowType: 'sequence' 
              }
          }
      });
    },
    validateConnection: function(cellViewS, _magnetS, cellViewT, _magnetT, _end) {
      var source = cellViewS.model;
      var target = cellViewT.model;
      // don't allow loop links
      if (source === target) return false;
      // don't allow link to link connection
      if (source.isLink()) return false;
      if (target.isLink()) return false;
      // don't allow group connections
      var sourceType = source.get('type');
      var targetType = target.get('type');
      if (sourceType === BPMNTypes.Group || targetType === BPMNTypes.Group) return false;
      if (sourceType === BPMNTypes.Pool || targetType === BPMNTypes.Pool) return false;
      return true;
      },
      defaultAnchor: {
          name: 'perpendicular'
      },
      defaultConnectionPoint: {
          name: 'boundary',
          args: { stroke: true }
      },
      // Embedding
      embeddingMode: true,
      frontParentOnly: false,
      validateEmbedding: function(childView, parentView) {
          var parentType = parentView.model.get('type');
          var childType = childView.model.get('type');
          if (parentType === BPMNTypes.Pool && childType !== BPMNTypes.Pool) return true;
          if (parentType === BPMNTypes.Activity && childType === BPMNTypes.Event) return true;
          if (parentType === BPMNTypes.Group && childType !== BPMNTypes.Group && childType !== BPMNTypes.Pool) return true;
          return false;
      },
      // Highlighting
      highlighting: {
          default: {
              name: 'mask',
              options: {
                  attrs: {
                      'stroke': '#3498db',
                      'stroke-width': 3,
                      'stroke-linejoin': 'round'
                  }
              }
          }
      }
  }).on({
  
      'blank:pointerdown': function(evt, x, y) {
          closeTools();
          selection.startSelecting(evt);
      },
  
      'blank:contextmenu': function(evt, x, y) {
          paperScroller.startPanning(evt);
      },
  
      'cell:pointerclick': function(cellView, evt, x, y) {
          openTools(cellView);
      },
  
      'link:mouseenter': function(linkView) {
          // Open tool only if there is none yet
          if (linkView.hasTools()) return;
  
          var ns = linkTools;
          var toolsView = new dia.ToolsView({
              name: 'link-hover',
              tools: [
                  new ns.Vertices({ vertexAdding: false }),
                  new ns.SourceArrowhead(),
                  new ns.TargetArrowhead()
              ]
          });
  
          linkView.addTools(toolsView);
      },
  
      'link:mouseleave': function(linkView) {
          // Remove only the hover tool, not the pointerdown tool
          if (linkView.hasTools('link-hover')) {
              linkView.removeTools();
          }
      },
  
      'link:connect': function(linkView) {
          // Change the link type based on the connected elements
          var link = linkView.model;
          var source = link.getSourceCell();
          var target = link.getTargetCell();
          if (!source || !target) return;
          var types = [source.get('type'), target.get('type')];
          var linkType = link.get('type');
          if (types.indexOf(BPMNTypes.Annotation) > -1) {
              if (linkType === BPMNTypes.AnnotationLink) return;
              replaceLink(graph, link, BPMNTypes.AnnotationLink);
              return;
          }
          if (types.indexOf(BPMNTypes.Conversation) > -1) {
              if (linkType === BPMNTypes.ConversationLink) return;
              replaceLink(graph, link, BPMNTypes.ConversationLink);
              return;
          }
          if (types.indexOf(BPMNTypes.DataObject) > -1) {
              if (linkType === BPMNTypes.DataAssociation) return;
              replaceLink(graph, link, BPMNTypes.DataAssociation);
              return;
          }
          if (types.indexOf(BPMNTypes.DataStore) > -1) {
              if (linkType === BPMNTypes.DataAssociation) return;
              replaceLink(graph, link, BPMNTypes.DataAssociation);
              return;
          }
          if (linkType !== BPMNTypes.Flow) {
              replaceLink(graph, link, BPMNTypes.Flow);
              return;
          }
      },
  
      'paper:pan': function(evt, tx, ty) {
          evt.preventDefault();
          paperScroller.el.scrollLeft += tx;
          paperScroller.el.scrollTop += ty;
      },
  
      'paper:pinch': function(evt, ox, oy, scale) {
          // the default is already prevented
          paperScroller.zoom(paperScroller.zoom() * scale, { min: 0.2, max: 5, ox, oy, absolute: true });
      }
  
  });
  
  var paperScroller = new ui.PaperScroller({
    autoResizePaper: true,
    padding: 0,
    paper: paper,
    scrollWhileDragging: true
  });
  
  const hasPaperClass = toolbarContainer.current.classList.contains('paper-container');
  if(!hasPaperClass){
    paperContainer.current.appendChild(paperScroller.el);
    toolbarContainer.current.classList.add('paper-container');
  }
  paperScroller.center();

  var selection = new ui.Selection({
    paper: paper,
    graph: graph,
    useModelGeometry: true,
    filter: [BPMNTypes.Pool, BPMNTypes.Group] // don't allow to select pool or group shapes
  });
  
  var stencil = new ui.Stencil({
    graph: graph,
    paper: paperScroller,
    width: 1000,
    height: 100,
    dragEndClone: function(cell) {
  
        var clone:any = cell.clone();
        var type = clone.get('type');
  
        // some types of the elements need resizing after they are dropped
        var sizeMultipliers: any = {};
        sizeMultipliers[BPMNTypes.Pool] = 6;
        sizeMultipliers[BPMNTypes.Choreography] = 2;
        sizeMultipliers[BPMNTypes.Group] = 2;
  
        if (type in sizeMultipliers) {
            var multiplier = sizeMultipliers[type]
            var originalSize = clone.size();
            clone.set('size', {
                width: originalSize.width * multiplier,
                height: originalSize.height * multiplier
            });
        }
  
        if (type === BPMNTypes.Pool) {
            var height = clone.size().height;
            clone.set({
                lanes: [{
                    label: 'Lane 1',
                    size: height / 2
                }, {
                    label: 'Lane 2',
                    size: height / 2
                }],
                padding: { top: 0, left: 30, right: 0, bottom: 0 }
            });
        }
        return clone;
    }
  });
  
  if (stencilContainer.current) {
    const hasMyClass = stencilContainer.current.classList.contains('stencil-container');
    stencil.render();
    if(!hasMyClass){
        stencilContainer.current.appendChild(stencil.el);
        stencilContainer.current.classList.add('stencil-container');
    }
  }
  stencil.load([{
    type: BPMNTypes.Activity,
    size: {
      width: 50,
      height: 50
    }
  }, {
    type: BPMNTypes.Gateway
  }, {
    type: BPMNTypes.Event
  }, {
    type: BPMNTypes.Conversation
  }, {
    type: BPMNTypes.DataAssociation
  }, {
    type: BPMNTypes.DataStore
  }, {
    type: BPMNTypes.DataObject
  }, {
    type: BPMNTypes.Annotation,
    attrs: {
      body: {
        fill: '#ffffff'
      }
    }
  }, {
    type: BPMNTypes.Group,
    size: {
        width: 50,
        height: 50
    },
    attrs: {
        label: { text: 'Group' }
    }
  }, {
    type: BPMNTypes.Pool,
    padding: { top: 0, left: 15, right: 0, bottom: 0 },
    size: { width: 90, height: 50 },
    lanes: [{
        label: 'Lane 1'
    }, {
        label: 'Lane 2'
    }],
    attrs: {
        body: { fill: '#ffffff' },
        headerLabel: { text: 'Header' },
        laneLabels: { fontSize: 13 },
        milestoneLabels: { fontSize: 13 }
    }
  }, {
    type: BPMNTypes.Choreography,
    participants: ['Participant 1', 'Participant 2']
  }]);
  layout.GridLayout.layout(stencil.getGraph(), {
    columns: 100,
    columnWidth: 'compact',
    marginX: 10,
    marginY: 10,
    columnGap: 20,
    // verticalGap: 20,
    resizeToFit: false,
  });
  stencil.on('element:drop', function(elementView, _evt, _x, _y) {
    // open inspector after a new element dropped from stencil
    openTools(elementView);

    setDroppedElements((prevElements:any) => [
      ...prevElements,
      {value:elementView.model.get('type'), fullEle: elementView} // You can add whatever information you want here
  ]);
  });
  
  /* KEYBOARD */
  
  keyboard.on('delete backspace', function() {
    graph.removeCells(selection.collection.toArray());
  });
  
  /* TOOLBAR */
  
  var toolbar = new ui.Toolbar({
    tools: toolbarConfig.tools,
    references: {
        paperScroller: paperScroller,
        commandManager: commandManager
    }
  });
  
  toolbar.on({
    'to-json:pointerclick': function() {
        var windowFeatures = 'menubar=no,location=no,resizable=yes,scrollbars=yes,status=no';
        var windowName = _.uniqueId('json_output');
        var jsonWindow = window.open('', windowName, windowFeatures);
        if (jsonWindow) {
            jsonWindow.document.write(JSON.stringify(graph.toJSON()));
        }
    },
    'clear:pointerclick': function() {
        graph.clear();
        paperScroller.center(500, 500);
    },
    'print:pointerclick': function() {
        paper.print();
    }
  });
  
  
  const hasMyClass = toolbarContainer.current.classList.contains('toolbar-container');
  toolbar.render();
  if(!hasMyClass){
    toolbarContainer.current.appendChild(toolbar.el);
    toolbarContainer.current.classList.add('toolbar-container');
  }
  /* TOOLTIPS */
  new ui.Tooltip({
    target: '[data-tooltip]',
    content: function(el:any) { return el.dataset.tooltip; },
    top: '.joint-toolbar',
    padding: 10,
    // direction: 'top'
  });
  
  // Create tooltips for all the shapes in stencil.
  stencil.getGraph().getElements().forEach(function(cell) {
    new ui.Tooltip({
      target: '.joint-stencil [model-id="' + cell.id + '"]',
      content: cell.get('type').split('.')[1],
      bottom: '.joint-stencil',
      // direction: 'bottom',
      padding: 0
    });
  });
  // load an example graph
  graph.fromJSON(example);
  paperScroller.positionContent('top-left', { padding: 50, useModelGeometry: true });
  /* ACTIONS */
  
  }
  
  function closeTools() {
    paper.removeTools();
    ui.Inspector.close();
    ui.FreeTransform.clear(paper);
    ui.Halo.clear(paper);
  }
  
  function openTools(cellView: any, coordinates: any = null) {
    closeTools();
    var cell = cellView.model;
    var type = cell.get('type');
    selection.collection.reset([]);
    // Add the cell into the selection collection silently
    // so no selection box is rendered above the cellView.
    selection.collection.add(cell, { silent: true });
    if (cell.isElement()) {
      createElementHalo(cellView);
      createElementFreeTransform(cellView);
        if (type === BPMNTypes.Pool && coordinates) {
          createPoolTools(cellView, coordinates);
        }
    } else {
      createLinkTools(cellView);
    }
    ui.Inspector.create('#inspector-container', {
      cell: cell,
      inputs: inputs[type],
      groups: {
        general: { label: type, index: 1 },
        appearance: { index: 2 },
        defaults: { index: 3 }
      }
    });
  }
  
  function createPoolTools(poolView: any, coordinates: any) {
    var pool = poolView.model;
    // If there is a lane under the pointer (mouse/touch),
    // add the swimlane tools and remove the FreeTransform from the cell
    var lanesIds = pool.getLanesFromPoint(coordinates);
    if (lanesIds.length === 0) return;
    var laneId = lanesIds[0];
    var boundaryTool = new elementTools.SwimlaneBoundary({
      laneId: laneId,
      padding: 0,
      attributes: {
        'fill': 'none',
        'stroke-width': 3,
        'stroke': '#3498db'
      }
    });
    var transformTool = new elementTools.SwimlaneTransform({
        laneId: laneId,
        minSize: 60,
        padding: 0,
    });
    var elementToolsView = new dia.ToolsView({
        tools: [boundaryTool, transformTool]
    });
    poolView.addTools(elementToolsView);
    ui.FreeTransform.clear(paper);
  }
  
  function createLinkTools(linkView: any) {
    var ns = linkTools;
    var toolsView = new dia.ToolsView({
      name: 'link-pointerdown',
      tools: [
        new ns.Vertices(),
        new ns.SourceAnchor(),
        new ns.TargetAnchor(),
        new ns.SourceArrowhead(),
        new ns.TargetArrowhead(),
        new ns.Segments(),
        new ns.Boundary({ padding: 15 }),
        new ns.Remove({ offset: -20, distance: 40 })
      ]
    });
    linkView.addTools(toolsView);
  }

  function createElementHalo(cellView: any) {
    var type = cellView.model.get('type');
    var halo = new ui.Halo({
        cellView: cellView,
        theme: 'default',
        type: 'toolbar',
        useModelGeometry: true,
        boxContent: type
    });
    halo.removeHandle('rotate');
    halo.removeHandle('resize');
    if (type === BPMNTypes.Pool || type === BPMNTypes.Group) {
        halo.removeHandle('link');
        halo.removeHandle('fork');
        halo.removeHandle('unlink');
    }
    halo.render();
  }
  
    function createElementFreeTransform(cellView: any) {
      var defaultMinSize = 30;
      var freeTransform = new ui.FreeTransform({
          cellView: cellView,
          allowOrthogonalResize: false,
          allowRotation: false,
          minWidth: function(el:any) {
              return (el.get('type') === BPMNTypes.Pool) ? el.getMinimalSize().width : defaultMinSize;
          },
          minHeight: function(el:any) {
              return (el.get('type') === BPMNTypes.Pool) ? el.getMinimalSize().height : defaultMinSize;
          }
      });
      freeTransform.render();
    }
  
    function replaceLink(linkGraph: any, link: any, type: any) {
      var link2JSON = {
        type: type,
        source: link.source(),
        target: link.target(),
        vertices: link.vertices()
      };
      link.remove();
      linkGraph.addCell(link2JSON);
    }
  
  }, []);
  
  return (
    <>
      <div id="paper-container" ref={paperContainer}></div>
      <div id="toolbar-container" ref={toolbarContainer}></div>
      <div id="stencil-container" ref={stencilContainer}></div>
      <div id="inspector-container">
        <div className="about">
          <h1>Business Process Model and Notation</h1>
        </div>
      </div>
      <ul>
                        {droppedElements.map((element:any, index:any) => (
                            <li key={index} onClick={()=>getEl(element.fullEle)}>{element.value}</li>
                        ))}
                    </ul>
    </>
  )
}
```

export default Bpmne

**Comment by kumilingus** [comment]
Duplicate of https://github.com/clientIO/joint/discussions/2472?

**Reply by nikita-fxbytes** [reply]
**i want it like this**
1. The header should contain buttons for all shapes. On clicking the shape that shape will appear on the paper with ui.Halo

2. Whichever shape I click on, that shape appears in the side bar with its text.

3. Whichever shape I click on, its ui.Inspector appears.

4. If I drag and drop the shape on the paper, the position of the shape will change in the side bar and if any shape has any connection then it will also appear on the side as if it were a child or parent.

5. I can edit, delete shapes

How will I do all this in joint form?

my code i know it has something wrong
What correction should I make in the code?


'use client'
import React, { useEffect, useRef, useState } from 'react';
import { dia, ui, shapes, elementTools,linkTools, layout } from '@clientio/rappid';
import _ from 'lodash';
import { toolList } from './toolbar';
import { examples } from './example';
import { inputss } from './inspector';
const Bpmne = () => {
    const toolbarContainer: React.RefObject<HTMLDivElement> = useRef(null);
    const paperContainer: React.RefObject<HTMLDivElement> = useRef(null);


    const [droppedElements, setDroppedElements] = useState<any>([]); // State to hold dropped elements

    // New
    // End

    // var graph = new dia.Graph({ type: 'bpmn' }, { cellNamespace: shapes });
    const [graph] = useState(new dia.Graph({ type: 'bpmn'}, { cellNamespace: shapes }));

    let dataD: any = null

    const getEl = (e:any) => {
      var inputs:any = inputss;
      var cell = e;
      var type = cell.get('type');
      ui.Inspector.create('#inspector-container', {
        cell: cell,
        inputs: inputs[type],
        groups: {
          general: { label: type, index: 1 },
          appearance: { index: 2 },
          defaults: { index: 3 }
        }
      });
    }
  useEffect(() => {
    setDroppedElements(graph.toJSON().cells)
    console.log(graph.toJSON().cells, 'graph.toJSON().cells')
    if (
      paperContainer.current &&
      toolbarContainer.current
    ) {
    var example = examples;
    var inputs:any = inputss;
    var toolbarConfig = toolList;
    var bpmn2:any = shapes.bpmn2;
    var BPMNTypes = {
      Pool: 'bpmn2.HeaderedPool',
      Group: 'bpmn2.Group',
      Activity: 'bpmn2.Activity',
      Event: 'bpmn2.Event',
      Gateway: 'bpmn2.Gateway',
      DataObject: 'bpmn2.DataObject',
      DataStore:'bpmn2.DataStore',
      DataAssociation: 'bpmn2.DataAssociation',
      Flow: 'bpmn2.Flow',
      Conversation: 'bpmn2.Conversation',
      ConversationLink: 'bpmn2.ConversationLink',
      Annotation: 'bpmn2.Annotation',
      AnnotationLink: 'bpmn2.AnnotationLink',
      // Legacy BPMN Shape
      Choreography: 'bpmn.Choreography'
    };
    // var graph = new dia.Graph({ type: 'bpmn' }, { cellNamespace: shapes });
    var commandManager = new dia.CommandManager({ graph: graph });
    var keyboard = new ui.Keyboard();
  
  
    var paper = new dia.Paper({
      width: 600,
      height: 100,
      gridSize: 1,
      model: graph,
      async: true,
      sorting: dia.Paper.sorting.APPROX,
      interactive: { linkMove: false },
      snapLabels: true,
      cellViewNamespace: shapes,
      clickThreshold: 10,
      // Connections
      defaultLink: function () {
        return new bpmn2.Flow({
          attrs: {
              line: {
                  flowType: 'sequence' 
              }
          }
      });
    },
    validateConnection: function(cellViewS, _magnetS, cellViewT, _magnetT, _end) {
      var source = cellViewS.model;
      var target = cellViewT.model;
      // don't allow loop links
      if (source === target) return false;
      // don't allow link to link connection
      if (source.isLink()) return false;
      if (target.isLink()) return false;
      // don't allow group connections
      var sourceType = source.get('type');
      var targetType = target.get('type');
      if (sourceType === BPMNTypes.Group || targetType === BPMNTypes.Group) return false;
      if (sourceType === BPMNTypes.Pool || targetType === BPMNTypes.Pool) return false;
      return true;
      },
      defaultAnchor: {
          name: 'perpendicular'
      },
      defaultConnectionPoint: {
          name: 'boundary',
          args: { stroke: true }
      },
      // Embedding
      embeddingMode: true,
      frontParentOnly: false,
      validateEmbedding: function(childView, parentView) {
          var parentType = parentView.model.get('type');
          var childType = childView.model.get('type');
          if (parentType === BPMNTypes.Pool && childType !== BPMNTypes.Pool) return true;
          if (parentType === BPMNTypes.Activity && childType === BPMNTypes.Event) return true;
          if (parentType === BPMNTypes.Group && childType !== BPMNTypes.Group && childType !== BPMNTypes.Pool) return true;
          return false;
      },
      // Highlighting
      highlighting: {
          default: {
              name: 'mask',
              options: {
                  attrs: {
                      'stroke': '#3498db',
                      'stroke-width': 3,
                      'stroke-linejoin': 'round'
                  }
              }
          }
      }
  }).on({
      
      'blank:pointerdown': function(evt, x, y) {
        console.log('pointerdown')
        setDroppedElements(graph.toJSON().cells)
          closeTools();
          selection.startSelecting(evt);
      },
     
      'element:pointerup': function(evt, x, y) {
        setDroppedElements(graph.toJSON().cells)
        console.log('first')
    },
      'blank:contextmenu': function(evt, x, y) {
          paperScroller.startPanning(evt);
      },
  
      'cell:pointerclick': function(cellView, evt, x, y) {
        dataD = cellView.model
          openTools(cellView);
          graph.on('remove', function() {
            setDroppedElements(graph.toJSON().cells)
        });
      },
  
      'link:mouseenter': function(linkView) {
          // Open tool only if there is none yet
          if (linkView.hasTools()) return;
  
          var ns = linkTools;
          var toolsView = new dia.ToolsView({
              name: 'link-hover',
              tools: [
                  new ns.Vertices({ vertexAdding: false }),
                  new ns.SourceArrowhead(),
                  new ns.TargetArrowhead()
              ]
          });
  
          linkView.addTools(toolsView);
      },
  
      'link:mouseleave': function(linkView) {
          // Remove only the hover tool, not the pointerdown tool
          if (linkView.hasTools('link-hover')) {
              linkView.removeTools();
          }
      },
  
      'link:connect': function(linkView) {
          // Change the link type based on the connected elements
          var link = linkView.model;
          var source = link.getSourceCell();
          var target = link.getTargetCell();
          if (!source || !target) return;
          var types = [source.get('type'), target.get('type')];
          var linkType = link.get('type');
          if (types.indexOf(BPMNTypes.Annotation) > -1) {
              if (linkType === BPMNTypes.AnnotationLink) return;
              replaceLink(graph, link, BPMNTypes.AnnotationLink);
              return;
          }
          if (types.indexOf(BPMNTypes.Conversation) > -1) {
              if (linkType === BPMNTypes.ConversationLink) return;
              replaceLink(graph, link, BPMNTypes.ConversationLink);
              return;
          }
          if (types.indexOf(BPMNTypes.DataObject) > -1) {
              if (linkType === BPMNTypes.DataAssociation) return;
              replaceLink(graph, link, BPMNTypes.DataAssociation);
              return;
          }
          if (types.indexOf(BPMNTypes.DataStore) > -1) {
              if (linkType === BPMNTypes.DataAssociation) return;
              replaceLink(graph, link, BPMNTypes.DataAssociation);
              return;
          }
          if (linkType !== BPMNTypes.Flow) {
              replaceLink(graph, link, BPMNTypes.Flow);
              return;
          }
      },
  
      'paper:pan': function(evt, tx, ty) {
          evt.preventDefault();
          paperScroller.el.scrollLeft += tx;
          paperScroller.el.scrollTop += ty;
      },
  
      'paper:pinch': function(evt, ox, oy, scale) {
          // the default is already prevented
          paperScroller.zoom(paperScroller.zoom() * scale, { min: 0.2, max: 5, ox, oy, absolute: true });
      }
  
  });
  
  var paperScroller = new ui.PaperScroller({
    autoResizePaper: true,
    padding: 0,
    paper: paper,
    scrollWhileDragging: true
  });
  
  const hasPaperClass = paperContainer.current.classList.contains('paper-container');
  if(!hasPaperClass){
    paperContainer.current.appendChild(paperScroller.el);
    paperContainer.current.classList.add('paper-container');
  }
  paperScroller.center();

  var selection = new ui.Selection({
    paper: paper,
    graph: graph,
    useModelGeometry: true,
    filter: [BPMNTypes.Pool, BPMNTypes.Group] // don't allow to select pool or group shapes
  });
  


  
  
  /* KEYBOARD */
  
  keyboard.on('delete backspace', function() {
    graph.removeCells(selection.collection.toArray());
  });
  
  /* TOOLBAR */
  
  var toolbar = new ui.Toolbar({
    tools: toolbarConfig.tools,
    references: {
        paperScroller: paperScroller,
        commandManager: commandManager
    }
  });
  
  toolbar.on({
    'to-json:pointerclick': function() {
        var windowFeatures = 'menubar=no,location=no,resizable=yes,scrollbars=yes,status=no';
        var windowName = _.uniqueId('json_output');
        var jsonWindow = window.open('', windowName, windowFeatures);
        if (jsonWindow) {
            jsonWindow.document.write(JSON.stringify(graph.toJSON()));
        }
    },
    'clear:pointerclick': function() {
        graph.clear();
        paperScroller.center(500, 500);
        setDroppedElements(graph.toJSON().cells)
    },
    'print:pointerclick': function() {
        paper.print();
    }
  });
  
  
  const hasMyClass = toolbarContainer.current.classList.contains('toolbar-container');
  toolbar.render();
  if(!hasMyClass){
    toolbarContainer.current.appendChild(toolbar.el);
    toolbarContainer.current.classList.add('toolbar-container');
  }
  /* TOOLTIPS */
  new ui.Tooltip({
    target: '[data-tooltip]',
    content: function(el:any) { return el.dataset.tooltip; },
    top: '.joint-toolbar',
    padding: 10,
    // direction: 'top'
  });
  
  // Create tooltips for all the shapes in stencil.

  // load an example graph
  graph.fromJSON(example);
  paperScroller.positionContent('top-left', { padding: 50, useModelGeometry: true });
  /* ACTIONS */
  
  }
  
  function closeTools() {
    paper.removeTools();
    ui.Inspector.close();
    ui.FreeTransform.clear(paper);
    ui.Halo.clear(paper);
  }
  
  function openTools(cellView: any, coordinates: any = null) {
    closeTools();
    var cell = cellView.model;
    var type = cell.get('type');
    selection.collection.reset([]);
    // Add the cell into the selection collection silently
    // so no selection box is rendered above the cellView.
    selection.collection.add(cell, { silent: true });
    if (cell.isElement()) {
      createElementHalo(cellView);
      createElementFreeTransform(cellView);
        if (type === BPMNTypes.Pool && coordinates) {
          createPoolTools(cellView, coordinates);
        }
    } else {
      createLinkTools(cellView);
    }
    ui.Inspector.create('#inspector-container', {
      cell: cell,
      inputs: inputs[type],
      groups: {
        general: { label: type, index: 1 },
        appearance: { index: 2 },
        defaults: { index: 3 }
      }
    });
  }
  
  function createPoolTools(poolView: any, coordinates: any) {
    var pool = poolView.model;
    // If there is a lane under the pointer (mouse/touch),
    // add the swimlane tools and remove the FreeTransform from the cell
    var lanesIds = pool.getLanesFromPoint(coordinates);
    if (lanesIds.length === 0) return;
    var laneId = lanesIds[0];
    var boundaryTool = new elementTools.SwimlaneBoundary({
      laneId: laneId,
      padding: 0,
      attributes: {
        'fill': 'none',
        'stroke-width': 3,
        'stroke': '#3498db'
      }
    });
    var transformTool = new elementTools.SwimlaneTransform({
        laneId: laneId,
        minSize: 60,
        padding: 0,
    });
    var elementToolsView = new dia.ToolsView({
        tools: [boundaryTool, transformTool]
    });
    poolView.addTools(elementToolsView);
    ui.FreeTransform.clear(paper);
  }
  
  function createLinkTools(linkView: any) {
    var ns = linkTools;
    var toolsView = new dia.ToolsView({
      name: 'link-pointerdown',
      tools: [
        new ns.Vertices(),
        new ns.SourceAnchor(),
        new ns.TargetAnchor(),
        new ns.SourceArrowhead(),
        new ns.TargetArrowhead(),
        new ns.Segments(),
        new ns.Boundary({ padding: 15 }),
        new ns.Remove({ offset: -20, distance: 40 })
      ]
    });
    linkView.addTools(toolsView);
  }

  function createElementHalo(cellView: any) {
    var type = cellView.model.get('type');
    var halo = new ui.Halo({
        cellView: cellView,
        theme: 'default',
        type: 'toolbar',
        useModelGeometry: true,
        boxContent: type
    });
    halo.removeHandle('rotate');
    halo.removeHandle('resize');
    if (type === BPMNTypes.Pool || type === BPMNTypes.Group) {
        halo.removeHandle('link');
        halo.removeHandle('fork');
        halo.removeHandle('unlink');
    }
    halo.render();
  }
  
    function createElementFreeTransform(cellView: any) {
      var defaultMinSize = 30;
      var freeTransform = new ui.FreeTransform({
          cellView: cellView,
          allowOrthogonalResize: false,
          allowRotation: false,
          minWidth: function(el:any) {
              return (el.get('type') === BPMNTypes.Pool) ? el.getMinimalSize().width : defaultMinSize;
          },
          minHeight: function(el:any) {
              return (el.get('type') === BPMNTypes.Pool) ? el.getMinimalSize().height : defaultMinSize;
          }
      });
      freeTransform.render();
    }
  
    function replaceLink(linkGraph: any, link: any, type: any) {
      var link2JSON = {
        type: type,
        source: link.source(),
        target: link.target(),
        vertices: link.vertices()
      };
      link.remove();
      linkGraph.addCell(link2JSON);
    }
  
  }, []);
 
  const add = (data:any) => {
    const x = dataD && dataD.attributes && dataD.attributes.position && dataD.attributes.position.x ? dataD.attributes.position.x + (Math.random() * 100) : Math.random() * 800
    const y = dataD && dataD.attributes && dataD.attributes.position && dataD.attributes.position.y ? dataD.attributes?.position.y :  300
    var rect:any = ''
    if(data === 'Activity'){
       rect = new shapes.bpmn2.Activity({
        position: { x: x, y: y },
        attrs: {
          'label': { text: 'My Activity' },
          'markers': {
              iconTypes: ['ad-hoc', 'loop'],
              iconColor: '#ffffff'
          }
      }
       });
    }
    if(data === 'Gateway'){
       rect = new shapes.bpmn2.Gateway();
    }
    rect.addTo(graph);
    var inputs:any = inputss;
    var cell =rect;
    var type = rect.attributes.type;
    ui.Inspector.create('#inspector-container', {
      cell: cell,
      inputs: inputs[type],
      groups: {
        general: { label: type, index: 1 },
        appearance: { index: 2 },
        defaults: { index: 3 }
      }
    });
    setDroppedElements(graph.toJSON().cells)
    // setDroppedElements((prevElements:any) => [
    //   ...prevElements,
    //   {rect} // You can add whatever information you want here
    // ]);


    
  }
  
  return (
    <>
      <div id="paper-container" ref={paperContainer}></div>
      <div id="toolbar-container" ref={toolbarContainer}></div>
      <div id="stencil-container">
        <button onClick={()=>add('Activity')}>Activity</button>
        <button onClick={()=>add('Gateway')}>Gateway</button>
      </div>
      <div id="inspector-container">
        <div className="about">
          <h1>Business Process Model and Notation</h1>
        </div>
      </div>
     

      <ul>
                        {droppedElements.map((element:any, index:any) => (
                            <li key={index} onClick={()=>getEl(element)}>{element.type}</li>
                        ))}
                    </ul>
    </>
  )
}

export default Bpmne

my UI is this type

![image_2024_01_30T05_40_14_455Z](https://github.com/clientIO/joint/assets/99159940/6e3bafda-bf2e-461d-910e-188160c77780)

**Reply by kumilingus** [reply]
Please check out our guide on how to write good technical questions: https://github.com/clientIO/joint/wiki/Asking-Technical-Questions.

My main suggestion would be to try dividing this question into multiple smaller questions.

**Comment by kumilingus** [comment]
Duplicate of https://github.com/clientIO/joint/discussions/2513.
